service/debugger.go

```diff
@@ -0,0 +1,402 @@
+package service
+
+import (
+	"fmt"
+	"regexp"
+	"runtime"
+
+	"github.com/golang/glog"
+	sys "golang.org/x/sys/unix"
+
+	"github.com/derekparker/delve/proctl"
+	"github.com/derekparker/delve/service/api"
+)
+
+// Debugger provides a thread-safe DebuggedProcess service. Instances of
+// Debugger can be exposed by other services.
+type Debugger struct {
+	processFuncs chan func(*proctl.DebuggedProcess) error
+	stop         chan bool
+}
+
+func NewDebugger(stop chan bool) *Debugger {
+	debugger := &Debugger{
+		processFuncs: make(chan func(*proctl.DebuggedProcess) error),
+		stop:         stop,
+	}
+	return debugger
+}
+
+// withProcess facilitates thread-safe access to the DebuggedProcess. All
+// interaction with DebuggedProcess should occur via calls to withProcess, and
+// the functions placed on the processFuncs channel should be consumed and
+// executed from the same thread as the DebuggedProcess.
+//
+// This is convenient because it allows things like HTTP handlers in
+// goroutines to work with the DebuggedProcess with synchronous semantics.
+func (d *Debugger) withProcess(f func(*proctl.DebuggedProcess) error) error {
+	done := make(chan interface{})
+	d.processFuncs <- func(proc *proctl.DebuggedProcess) error {
+		err := f(proc)
+		if err != nil {
+			done <- err
+		} else {
+			done <- struct{}{}
+		}
+		return nil
+	}
+	result := <-done
+	if e, ok := result.(error); ok {
+		return e
+	}
+	return nil
+}
+
+// TODO: Handle attaching
+func (d *Debugger) Run(processArgs []string) error {
+	// We must ensure here that we are running on the same thread during
+	// the execution of dbg. This is due to the fact that ptrace(2) expects
+	// all commands after PTRACE_ATTACH to come from the same thread.
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	glog.Infof("launching process with args: %v\n", processArgs)
+	process, err := proctl.Launch(processArgs)
```

Comment:
`Run`'s signature will have to be modified here to let us know if we are supposed to launch the process or attach to it.